### ISSUE TITLE
HA/FT Launcher

### ISSUE THREAD

**Issue by 1000TurquoisePogs** [zowe/zlux#590]
Using the [manifest specification](https://github.com/zowe/zowe-install-packaging/issues/1477) servers must provide scripts for lifecycle management and declare them in manifest files such that a Zowe tool "Zowe Launcher" can orchestrate the lifecycles of components and possibly their redundant copies for HA/FT.

This launcher should start requested components on STC startup, and stop components on Job cancelation, but by using signals such that the way in which the components are stopped is cross-platform and granular.

The launcher must monitor each started component to see if it stops before being requested to stop, such as a crash, and attempt to restart the server within user-defined limits of retry counts within a window of time to prevent looping.

The launcher should be designed with linux & z/os compatibility in mind, such that HA/FT for containers can be accomplished in the future. However, there may be z/os-specific features such as WTOR so that z/os users can be informed of when zowe servers have all started up successfully.

## Deliverable
A tool for managing highly-available servers by having multiple copies of Zowe components running, monitoring their status for crashes, and managing their lifecycle via component-provided scripts

## Dependency
People working on the launcher will need to cooperate with webui, apiml, system squads in order to ensure each component is working successfully with the launcher.

## Tasks
* Create a manifest file for app-server
* Create a manifest file for zss
* Build zlaunch in our pipeline
* Put zlaunch into convenience pax
* Zlaunch to discover each components lifecycle scripts at startup

**Comment by ifakhrutdinov** [comment]
@1000TurquoisePogs I would mention that termination command `P` must be handled as this is the biggest complain from the customers. Right now they have to actually cancel the started task. Here's the ticket which also links the other issues, https://github.com/zowe/zowe-install-packaging/issues/1544

**Comment by lchudinov** [comment]
### When Launcher is running as STC 

1. Launcher starts using INSTANCE_DIR provided as environment variable
2. Launcher executes a new script `save-environemnt.sh` which is copy of `run-zowe.sh` up to line https://github.com/zowe/zowe-install-packaging/blob/master/bin/internal/run-zowe.sh#L38 and contains a new line `env > ${INSTANCE_DIR}/saved-enviroment.env`
3. Launcher reads `${INSTANCE_DIR}/saved-enviroment.env` and restores environment
4. Launcher find components to start using `LAUNCH_COMPONENTS` environment variable
5. Launcher spawns each component as a child process using `start-component.sh` script
 
### When Launcher is running as a part of run-zowe.sh script 

1. Launcher starts using INSTANCE_DIR and ROOT_DIR provided as environment variables, and component list as a first command line argument
5. Launcher spawns each component as a child process using `start-component.sh` script

**Comment by ifakhrutdinov** [comment]
> Launcher reads ${INSTANCE_DIR}/saved-enviroment.env and restores environment

Why does the launcher process need to establish an environment? Could `start-component.sh` do that?

**Comment by lchudinov** [comment]
> Why does the launcher process need to establish an environment? Could start-component.sh do that?

We need to establish environment to find out which components to start (step4) and only after that start the components(step5).

**Comment by ifakhrutdinov** [comment]
Can we get a list of component from a config file?

**Comment by lchudinov** [comment]
We can get list of components by executing `bin/internal/get-launch-components.sh`  and reading `stdout`

**Comment by lchudinov** [comment]
### Related PRs:
- https://github.com/zowe/launcher/pull/18
- https://github.com/zowe/zowe-install-packaging/pull/1970

**Comment by 1000TurquoisePogs** [comment]
Currently, zowe can be started by STC or script, because the STC just runs the script anyway, so there is only 1 code path. This sounds like 2 code paths and it will be required anyway because docker cant use the STC code path. I'll take a look at the PRs to see how you're organizing that.

**Comment by 1000TurquoisePogs** [comment]
One thing I need clarification on:
Where will the launch control parameters be located?
Currently there are types of files:

1) $INSTANCE_DIR/instance.env, it contains user configuration
2) $ROOT_DIR/components/$component/manifest.yaml, it contains metadata about a component.

Did we already decide where parameters such as "how many times to restart a service within a period of time" will be located?
I think it would go into instance.env
Maybe components (via manifest.yaml) will have defaults. But something in $INSTANCE_DIR must be used to modify defaults.
So, zowe_launcher needs to be able to read those files.

**Comment by ifakhrutdinov** [comment]
The "read and restore" environment seems a bit complicated to me. Could we solve these issue by allowing components add their info to the launcher? For instance, install scripts could register components in some generic launcher config file and provide start scripts which would establish the required environment. All the launcher need to do is start those script. So basically what we are trying to do at runtime now would be solved at install time. That way the launcher wouldn't have to know so much details and wouldn't have to tied to zowe-install-packaging (and potentially changed every time there are changes there).

@lchudinov , @1000TurquoisePogs , @jackjia-ibm what do you think?

**Comment by lchudinov** [comment]
@ifakhrutdinov  As far as I can see, `https://github.com/zowe/zowe-install-packaging/blob/master/bin/internal/prepare-environment.sh` reads `instance.env` to determine the start component list. The script is not so easy to replace by a config file.

**Comment by ifakhrutdinov** [comment]
So, the biggest challenge is to figure out which components to start because that depends on the factors that aren't controlled by the launcher?

Suppose if we knew which components to start, could we call a wrapper script which would do all the heavy lifting, i.e. initializing the environment and then calling the actual start script?

**Comment by ifakhrutdinov** [comment]
Hi All, @lchudinov, @jackjia-ibm, @1000TurquoisePogs 

I think we're driving ourselves into a corner by putting zowe-install-packaging specific items into the launcher source code directly. The code is becoming much more fragile and inflexible.

Here's my proposal.

#### Launcher changes
* A new more flexible configuration file is created (perhaps YMAL?)
* The file contains a list of components (any components, not just Zowe) and their parameters (max retry count, same/separate address space, path to the binary, etc)

#### zowe-install-packaging changes

The zowe-install-packaging changes will be more involved, but this will help us in the long-run.

* Add a new generic wrapper script
  * The script will have variables like component name, root_dir, instance_dir, etc. which will be substituted by the deployment script (see below)
  * The script will be able to call a single component by reusing some refactored existing scripts

* Add a new deployment script which
  * Understand the launcher config format
  * Use the existing scripts to make a list of the components to deploy
  * Goes through the components and collects their launcher config if provided (e.g. a component may tell that it needs to be restarted if it crashes), it will be up to each component to provide its specific configuration
  * Generates a wrapper script for each component (substitutes the required paths and component name in the template)
  * Generates a launcher config (writes the component names, paths to their wrapper scripts, component parameters, etc.)
  * Injects the config into the launcher

Advantages:
* The launcher and packaging scripts are not coupled, each component (meaning the launcher and the rest of zowe-install-packaging) can be used completely independently
* If a Zowe customer wants to reuse the launcher for their application, they're not bound to anything in zowe-install-packaging, they can just use it by providing a simple config
* No need to deal with any environmental variables inside the launcher - the code is kept as simple as possible (as it should with C code)
* Anything that's needed by the Zowe components can be just "baked" into wrapper scripts, so you don't have to make any changes in the launcher code

Disadvantages:
* You'll need to re-run the deploy script when you want to apply new changes

Here's how I see the deployment process:

```
> cd zowe-install-packaging
> ./deploy-launcher.sh > launcher.yaml
```

Then we can either copy the config to the place where your launcher expects it, or we can provide a way to pass that info to the launcher at run-time (e.g. as an arg to the binary).

Let's have a call if needed. I think it's important to do this right so we don't keep having to change the launcher in the future when there are changes in the packaging/install scripts.

Thanks,
Irek

**Comment by 1000TurquoisePogs** [comment]
# Recap
What's missing so far is a step-by-step description of what happens when zowe starts, stops, and other actions.


This is what I mentioned last call:
## Idea 1
### Startup:
* Someone starts an STC or process, which runs a script to collects information about components
* The script eventually calls zowe launcher with information about which components to start and how to start them

### Shutdown:
* A signal is sent which reaches the parent script
* It is forwarded onto zowe launcher
* Launcher stops components either by sending an appropriate signal to their main process, or by calling a stop script if they provided one

### Configuration:
currently every component runs validate & configure before start. if configure needed to be done 10 times for 10 copies in HA/FT, then zowe launcher might need to be involved in the configure step. otherwise, configure can be done 1 time prior to running the zowe launcher.



Do my steps sound wrong? Because depending on how I read your comment, it sounds like:

## Idea 2
### Pre-startup, and any time modification is needed:
* Someone runs scripting to create wrapper scripts

### Startup:
* Someone starts an STC or process, which runs zowe launcher directly
* It reads a simple config about how to start each component

### Shutdown:
* A signal is sent to zowe launcher
* Launcher stops components either by sending an appropriate signal to their main process, or by calling a stop script if they provided one

### Configuration:
I guess the validate & config steps would be contained within the wrapper scripts. Basically the wrapper script would just be run-zowe.sh for 1 component

# Summary

It sounds like the best fit for what you describe is to have a file which describes which components to start and how to start them, but the "how to start them" is just a modification of "run-zowe.sh" which runs 1 component, such that run-zowe.sh is called multiple times.

This seems to have a few disadvantages
* Duplication of startup scripting leading to longer startup times (unsure if its slow enough to matter)
* Additional end-user steps needed to just run zowe

Anything that makes zowe more complicated to the end user is a big issue since it's a common pain point. 

There might be disadvantages of re-usability for the launcher by adding more knowledge of zowe into it, but it would be simpler to the end user. Can we accomplish both end-user simplicity and re-usability?

# Notes

>Launcher changes
>
>    A new more flexible configuration file is created (perhaps YMAL?)

While YAML-based configuration is already on the TODO list. Here's the summary:
* The yaml will have top-level attributes that correspond to a major section of zowe, such as app-server, zowe launcher, api gateway, and any products.
* Due to this, it can either be 1 file or multiple files that get merged into 1 file at runtime.

I don't know when that work will be complete, so in the short term keep every config necessary within instance.env
Because we will eventually be mapping existing instance.env content to the yaml config.
Less files is better, more files cause end-user confusion. Currently we have just 1 file, instance.env, so it's obvious where users can find config info.

>Add a new deployment script which
>Goes through the components and collects their launcher config if provided (e.g. a component may tell that it needs to be restarted if it crashes), it will be up to each component to provide its specific configuration

That config would be composed of defaults given in each component manifest.yml, and overrides given in the central config (currently instance.env)... its not up to each component exactly, its going to be a combination of defaults and end-user config if the defaults dont work.

>If a Zowe customer wants to reuse the launcher for their application, they're not bound to anything in zowe-install-packaging, they can just use it by providing a simple config

I think you're asking for zowe launcher to be a generic process-monitoring library that can be used unrelated to zowe.
And not that you want products to circumvent how zowe works, which is the sort of chaos zowe-install-packaging exists to prevent.

If the solution we make ends up with zowe launcher being a generic library, this is fine but it sounds incidental rather than a goal of Zowe.
Currently zowe launcher is a generic library, so if we change it and people don't like the change, I'd suggest they fork the old code. From what I understand of the licensing, people can do that without trouble and in particular you can since you own the code prior to the open-sourcing.

**Comment by ifakhrutdinov** [comment]
@1000TurquoisePogs 

> A signal is sent which reaches the parent script
> It is forwarded onto zowe launcher

Likely in the form of the termination operator command.

> This seems to have a few disadvantages
> Duplication of startup scripting leading to longer startup times (unsure if its slow enough to matter)

We don't have to do that. You can make the global validation a component which is run before every other component is run. If it fails it will stop the entire things. This could be indicated by priority in the config.

> Additional end-user steps needed to just run zowe

What do users do know to deploy Zowe right now? They probably run a whole lot of things (they run at least some script, don't they). This deployment step doesn't have to be run manually. It can be part of some other installation step. In fact, it must be part of some other step so we don't add more manual steps for users.

> I don't know when that work will be complete, so in the short term keep every config necessary within instance.env
Because we will eventually be mapping existing instance.env content to the yaml config.
> Less files is better, more files cause end-user confusion. Currently we have just 1 file, instance.env, so it's obvious where users can find config info.

The launcher config doesn't have to be edited/written by anyone, it should be a simple way to tell the launcher what to run. The installation/deployment will do that for users. These steps must be transparent.

> I think you're asking for zowe launcher to be a generic process-monitoring library that can be used unrelated to zowe.
And not that you want products to circumvent how zowe works, which is the sort of chaos zowe-install-packaging exists to prevent.

Of course I do! The more coupled something is, the harder it's to reuse. Imagine if someone wants to use this in a side project? This kind of stuff (STC handling) is a pain, and things like BPXBATCH aren't making things easier (because you can't even use `P` and `F` operator commands with them). We can then tell that person, that we have a generic component to solve such issues. I believe if people could re-use building blocks of Zowe, it'd only make it more popular. For instance, it's much more likely that someone uses our common library if it doesn't have any Zowe messages all over the code, and the more it's used the more contribution back to Zowe there is.

I understand that there is probably a schedule and pressure to release this launcher, so if things really need to be hardcoded into the C code, I'll be okay with that.

**Comment by jackjia-ibm** [comment]
Thanks for the feedbacks.

A little re-cap on how [run-zowe.sh](https://github.com/zowe/zowe-install-packaging/blob/staging/bin/internal/run-zowe.sh) currently works. It's separated into several independent scripts like this:

- `global-validate.sh`: anything validations we want to do before component level
- `prepare-workspace.sh`: it prepares environment and workspace folder, and also run all components validation/configure steps
- `get-launch-components.sh`: currently it returns a list of components should be started
- `start-component.sh`: start one component at a time. it will also call `prepare-environment.sh` again to configure environment variables, but not component validation/configure steps.

We already handle configure separately and is outside of `start-component.sh`. Seems the missing information is how we handle customizations for each component, like, if the components needs restart when it fails, how many times we can retry, if reuse zowe address space, etc. For now, will use a default value be a ok? Like we always try 10 restart when component fails with progressive internals (3x1s, 2x5s, 10s, 20s, 1m, etc), then output proper messages to log. So sysadm can investigate why it keeps failing. This might be part of benefit if we treat launcher like another component. In the past, we discussed component can define `configs` in `manifest.yaml`, and then there is parsed `.config-<ha-id>.yaml` be put into `workspace/<component>`. If it's a component, it inherit the behavior how we prepare configs for other components. But since launcher handles other components starting, the `commands.start` in manifest will not be supported. But `commands.install`, `commands.configureInstance` seems still make sense.

Currently the YAML config work I've done the instance.env to YAML format conversion, and working on changing the scripts to be compatible with the YAML config. But it won't catch up v1.20.0 release.

**Comment by 1000TurquoisePogs** [comment]
>>    This seems to have a few disadvantages
>>    Duplication of startup scripting leading to longer startup times (unsure if its slow enough to matter)

> We don't have to do that. You can make the global validation a component which is run before every other component is run. If it fails it will stop the entire things. This could be indicated by priority in the config.

That sounds good, but it might be complex to make the change.

>What do users do know to deploy Zowe right now?

There's 6 end-user actions today
1) Install: A script which puts files and datasets in places

2) Create instance: A script which creates a configuration area that is 1-to-many with the installation, so you can have multiple running zowe at once

3) Create keystore: A 1-to-many relationship with instances, it handles encryption keys

4) Modify instance: A single file, instance.env, is the user-facing configuration. Various files exist within an instance, but users shouldnt need to edit anything beyond instance.env or else someone failed at simplifying their config.

5) Modify keystore: A single file like instance.env, alongside scripts for automation, allow for keystore modification

6) Start: Either via JCL or single script within the instance directory. The job calls the script so its the same thing really.

So after the first run, they're down to just 2 actions, modify instance and start, which is just 1 file and 1 script.
If possible to have the start process perform automation prior to starting zowe launcher, then it might be possible to hide extra work from the zowe launcher by converting the instance config file contents into a usable form by zowe launcher prior to starting it.

**Comment by jackjia-ibm** [comment]
Want to write down few points we discussed in several meetings.

- Would like to suggest we separate enhancement from first version. I get the point and benefit to make it a generic start program and I totally agree. We can make it better once we know what's the best way. For now, to me it sounds we are not fully on same page about how Zowe should be started and may need more careful design. If we agree on releasing it for HA usage first, I believe we can merge https://github.com/zowe/launcher/pull/18 soon. If we can catch up 1.21.0 release, I don't mind we work on it early.
- Currently I think we don't need to start launcher form USS, with bin/internal/run-zowe.sh. Few reasons
  * It has chances to left subprocesses after launcher exits
  * It conflicts with the original interest of providing modify/F command for Zowe job
  * We may not need it when running in containers
- To run launcher as STC, would suggest we keep existing ZWESVSTC (to use BPXBATSL and execute bin/internal/run-zowe.sh), but add ZWELNCH as option 3 to start Zowe to keep backward compatibility. There are still many subtle differences if we replace ZWESVSTC  to execute ZWELNCH like how to terminate the job. Based on doc, option 1 is starting with <instance>/bin/zowe-start.sh, O=option 2 is using operator S ZWESVSTC command. Option 3 is the only way to support real HA.
- There is few more questions I have related to launcher to work with Zowe components:
  * If it's possible to remove the requirement for each component to trap INT signal in start.sh like this https://github.com/zowe/explorer-jes/blob/master/bin/start.sh#L25-L30
  * With the current way of start.sh, like above example, address spaces are doubled if we don't reuse/share address spaces. Is that possible to reduce it?
  * We will need to define proper log message IDs for major messages, like component started, failed, restarted, stop retrying, etc. Those message IDs will be useful for sysadmin to monitor log and take action.

**Comment by lchudinov** [comment]
@jackjia-ibm you had a suggestion about retry time intervals. Should we include it into the first version?
> Seems the missing information is how we handle customizations for each component, like, if the components needs restart when it fails, how many times we can retry, if reuse zowe address space, etc. For now, will use a default value be a ok? Like we always try 10 restart when component fails with progressive internals (3x1s, 2x5s, 10s, 20s, 1m, etc), then output proper messages to log. So sysadm can investigate why it keeps failing.

**Comment by Joe-Winchester** [comment]
On the HA call today we discussed whether each `start.sh` script could echo back details of the component it launched (the {JOB_NAME} together with its pid, the script exits, and the kill{ } loop is in the script that calls each `start.sh` and that is responsible for the trap and then iterates over all of the collected pids.  That way we wouldn't need each `start.sh` to stay around and it also means there'd be a way to centrally keep all of the process IDs.  If a `start.sh` launched more than one process with & it could return them all in an array that the overall script doing each component call to `start.sh` could collect.

**Comment by lchudinov** [comment]
> If a `start.sh` launched more than one process with & it could return them all in an array that the overall script doing each component call to `start.sh` could collect.

Knowing pids of each process helps to monitor failures, but the list of pids doesn't provide enough info for restarting the processes. To be able to restart each sub-process, launcher needs to know a start script for each of them, e.g.

```yml
commands:
  start: 
    - sub_component_a: bin/start_a.sh
    - sub_component_b: bin/start_b.sh
```

**Comment by Joe-Winchester** [comment]
given that the run script is calling the start.sh for each component, then couldn't it just re-call them if restart is needed ?  Trying to think of ways to avoid the start.sh script itself starting a child script so we don't double the address spaces, but leave the run_zowe.sh script as the sheepdog address space that is the parent for the start.sh scripts that just launch the node or Java address space and being started with BPXBATSL will terminate them all when the job is cancelled

**Comment by 1000TurquoisePogs** [comment]
Appears to be complete as of 1.23, but we can create bug tickets if needed.
